enterprise/backend/test/metabase_enterprise/dependencies/calculation_test.clj

```diff
@@ -383,3 +396,67 @@
+        (testing "dependent segment depends on both table and segment"
+          (is (= {:table #{products-id}
+                  :segment #{segment-a-id}}
+                 (calculation/upstream-deps:segment segment-b))))))))
+
+(deftest ^:parallel upstream-deps-card-with-multiple-segments-test
+  (testing "Card using multiple segments depends on all of them"
+    (let [products-id (mt/id :products)
+          price-field-id (mt/id :products :price)
+          category-field-id (mt/id :products :category)]
+      (mt/with-temp [:model/Segment {segment-a-id :id} {:table_id products-id
+                                                        :definition {:filter [:> [:field price-field-id nil] 50]}}
+                     :model/Segment {segment-b-id :id} {:table_id products-id
+                                                        :definition {:filter [:= [:field category-field-id nil] "Widget"]}}
+                     :model/Card card {:dataset_query {:database (mt/id)
+                                                       :type :query
+                                                       :query {:source-table products-id
+                                                               :filter [:and
+                                                                        [:segment segment-a-id]
+                                                                        [:segment segment-b-id]]}}}]
+        (is (= {:card #{}
+                :table #{products-id}
+                :segment #{segment-a-id segment-b-id}}
+               (calculation/upstream-deps:card card)))))))
+
+(deftest ^:parallel upstream-deps-segment-with-multiple-segments-test
+  (testing "Segment depending on multiple other segments tracks all dependencies"
+    (let [products-id (mt/id :products)
+          price-field-id (mt/id :products :price)
+          category-field-id (mt/id :products :category)
+          rating-field-id (mt/id :products :rating)]
+      (mt/with-temp [:model/Segment {segment-a-id :id} {:table_id products-id
+                                                        :definition {:filter [:> [:field price-field-id nil] 50]}}
+                     :model/Segment {segment-b-id :id} {:table_id products-id
+                                                        :definition {:filter [:= [:field category-field-id nil] "Widget"]}}
+                     :model/Segment segment-c {:table_id products-id
+                                               :definition {:filter [:and
+                                                                     [:segment segment-a-id]
+                                                                     [:segment segment-b-id]
+                                                                     [:> [:field rating-field-id nil] 4]]}}]
+        (is (= {:table #{products-id}
+                :segment #{segment-a-id segment-b-id}}
+               (calculation/upstream-deps:segment segment-c)))))))
```

Comment:
Missing a test case here: a segment that relies on an implicitly joined field should add a dep on that field's table.